File: src/libs/clientMessage.ts
"use strict"

export class ClientMessage {

    public method: string;
    public sender: string;
    public receiver: string;
    public code: string;
    public device: string;
    public roomid: string;
    public clientid: string;
    public extendParam: string;

    constructor() {
        this.roomid = '0';
        this.method = '';
        this.sender = '';
        this.receiver = '';
        this.code = '';
        this.device = '';
        this.clientid = '0';
        this.extendParam = '';
    }
}

export class IoTMessage {
    public method: string;
    public roomid: string;
    public clientid: number;
    public address: string;
    public rescode: string;
    public payload: any;

    constructor() {
        this.method = '';
        this.roomid = '';
        this.clientid = 0;
        this.address = '';
        this.rescode = '0';
        this.payload = undefined;
    }

    public static build(method: string, roomid: string, clientid: number, code : string, payload: any) : IoTMessage {
        let message = new IoTMessage();
        message.method = method ? method : '';
        message.roomid = roomid ? roomid : '';
        message.clientid = clientid ? clientid : 0;
        message.rescode = code ? code : '0';
        message.address = '';
        message.payload = payload ? payload : '';
        return message;
    }
}

export class ServerMessage {

    public message: any = '';
    public error: string = '';

    constructor(message?: any, error?: string) {
        this.message = message ?? "";
        this.error = error ?? "";
    }
}


// sendServerMsg sends a wsServerMsg composed from |msg| to the connection.
function sendServerMsg(ws: WebSocket, message: any) {
    if(ws.OPEN) {
        let m = new ServerMessage(message, '');
        send(ws, m);
    }
}

// sendServerErr sends a wsServerMsg composed from |errMsg| to the connection.
function sendServerErr(ws: WebSocket, error: string) {
    if(ws.OPEN) {
        let m = new ServerMessage('', error);
        send(ws, m);
    }
}

function send(ws: WebSocket, message: ServerMessage) {
    if(ws.OPEN) {
        let encodedString: string = JSON.stringify(message);
        ws.send(encodedString);
    }
}

export default {
    sendServerMsg,
    sendServerErr,
    send
};

File: services/websocket-relay/example/android/okhttp/models/ClientMessage.java
package com.wsrelay.api.models;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * RTC signalling message — the payload on /relay/rtc.
 *
 * Address format is "rtc:{address}@{host}", e.g. "rtc:101B203U@192.168.0.167".
 * The relay strips the "rtc:" prefix and the "@..." suffix when it looks up
 * push tokens, so the middle part must match the address you registered with
 * (POST /relay/register/mobile).
 *
 * Every field is a String, including roomid and clientid. The relay compares
 * them loosely but the server-side validator rejects messages where sender,
 * roomid or clientid is missing, so fill them in.
 */
public class ClientMessage {

    public String method = "";
    public String sender = "";
    public String receiver = "";
    public String code = "";
    public String device = "";
    public String roomid = "0";
    public String clientid = "0";

    /**
     * Free-form JSON string passed through untouched, with one exception:
     * on INVITE the relay parses it and, if it finds
     * {"targetAgent":"someone@example.com"}, sends the push only to that
     * account's devices instead of every device at the address.
     */
    public String extendParam = "";

    public ClientMessage() {}

    public ClientMessage(String method, String sender, String receiver, String roomid, String clientid) {
        this.method = method;
        this.sender = sender;
        this.receiver = receiver;
        this.roomid = roomid;
        this.clientid = clientid;
    }

    public static ClientMessage fromJson(JSONObject json) {
        ClientMessage m = new ClientMessage();
        m.method = json.optString("method", "");
        m.sender = json.optString("sender", "");
        m.receiver = json.optString("receiver", "");
        m.code = json.optString("code", "");
        m.device = json.optString("device", "");
        m.roomid = json.optString("roomid", "0");
        m.clientid = json.optString("clientid", "0");
        m.extendParam = json.optString("extendParam", "");
        return m;
    }

    public JSONObject toJson() throws JSONException {
        JSONObject o = new JSONObject();
        o.put("method", method);
        o.put("sender", sender);
        o.put("receiver", receiver);
        o.put("code", code);
        o.put("device", device);
        o.put("roomid", roomid);
        o.put("clientid", clientid);
        o.put("extendParam", extendParam == null ? "" : extendParam);
        return o;
    }

    /** Restrict an invite's push to one account. */
    public ClientMessage targetAgent(String email) throws JSONException {
        this.extendParam = new JSONObject().put("targetAgent", email).toString();
        return this;
    }

    /** Build "rtc:{address}@{host}". */
    public static String address(String address, String host) {
        return "rtc:" + address + "@" + host;
    }

    @Override
    public String toString() {
        return "ClientMessage{" + method + " room=" + roomid + " client=" + clientid
                + " " + sender + " -> " + receiver + "}";
    }
}
